hilti/include/rt/types/vector.h

```diff
@@ -75,98 +76,251 @@ bool operator!=(Allocator<T, D1> const&, Allocator<U, D2> const&) noexcept {
     return false;
 }
 
-template<typename T, typename Allocator = std::allocator<T>>
-class SafeIterator
-    : public hilti::rt::detail::iterator::SafeIterator<
-          Vector<T, Allocator>, typename Vector<T, Allocator>::SafeIterator, SafeIterator<T, Allocator>> {
+template<typename T, typename Allocator>
+class Iterator {
+    using V = Vector<T, Allocator>;
+
+    std::weak_ptr<Vector<T, Allocator>*> _control;
+    typename V::size_type _index = 0;
+
+    using reference = T&;
+
 public:
-    using Base =
-        hilti::rt::detail::iterator::SafeIterator<Vector<T, Allocator>, typename Vector<T, Allocator>::SafeIterator,
-                                                  SafeIterator<T, Allocator>>;
-    using Base::Base;
+    Iterator() = default;
+    Iterator(typename V::size_type&& index, const typename V::C& control)
+        : _control(control), _index(std::move(index)) {}
+
+    reference operator*();
+
+    Iterator& operator++() {
+        ++_index;
+        return *this;
+    }
+
+    Iterator operator++(int) {
+        auto ret = *this;
+        ++(*this);
+        return ret;
+    }
+
+    friend bool operator==(const Iterator& a, const Iterator& b) { return a._index == b._index; }
```

Comment:
Was wondering about this as well. Comparing iterators into different `std::vector` is strictly illegal, so it wouldn't implement that. Our situation is slightly different here as our iterators behave like indices (e.g., they remain valid beyond the lifetime of the iterated data). With that _not comparing `_control`_ is a valid and usable implementation (we'd compare indices in that case); comparing `_control` would just add as a sanity check that the user did only compare iterators into the same `Vector` though we wouldn't need that check for internal safety. WDYT?

Reply:
yeah, I think that sanity check would be good. In my mind an iterator is tied to a specific container, and it seems more intuitive to me to use that as the semantics for comparison as well so that two iterators pointing to different containers will never be considered equal. 

In fact, I can see going further and throw an exception if somebody even tries to compare iterators pointing to different containers because most likely it's evidence of a logic error somewhere. Let's do that I would say.

Reply:
I added validation that the iterators point to the same vector. I started writing our own control block, but it turns out I needed to allocate additional heap data at least once, so let's stay with `shared_ptr<Vector>` as control block for now. With that iterator comparisons will cause two read/write operation on the atomic use count inside the shared_ptr, but this should probably be fine. We can replace this implementation with a custom non-atomic shared_ptr implementation should this ever become an issue.